smarts/sstudio/genhistories.py

```diff
@@ -0,0 +1,446 @@
+               ) WITHOUT ROWID"""
+        )
+        ccur.execute(
+            """CREATE TABLE Vehicle (
+                   id INTEGER PRIMARY KEY,
+                   type INTEGER NOT NULL,
+                   length REAL,
+                   width REAL
+               ) WITHOUT ROWID"""
+        )
+        ccur.execute(
+            """CREATE TABLE Trajectory (
+                   vehicle_id INTEGER NOT NULL,
+                   sim_time REAL NOT NULL,
+                   position_x REAL NOT NULL,
+                   position_y REAL NOT NULL,
+                   heading_rad REAL NOT NULL,
+                   speed REAL DEFAULT 0.0,
+                   lane_id INTEGER DEFAULT 0,
+                   PRIMARY KEY (vehicle_id, sim_time),
+                   FOREIGN KEY (vehicle_id) REFERENCES Vehicles(id)
+               ) WITHOUT ROWID"""
+        )
+        dbconxn.commit()
+        ccur.close()
+
+    def create_output(self):
+        dbconxn = sqlite3.connect(self._output)
+
+        self._log.debug("creating tables...")
+        self._create_tables(dbconxn)
+
+        self._log.debug("inserting data...")
+
+        iscur = dbconxn.cursor()
+        insert_kv_sql = "INSERT INTO Spec VALUES (?, ?)"
+        self._write_dict(self._dataset_spec, insert_kv_sql, iscur)
+        dbconxn.commit()
+        iscur.close()
+
+        # TAI:  can use executemany() and batch insert rows together if this turns out to be too slow...
+        insert_vehicle_sql = "INSERT INTO Vehicle VALUES (?, ?, ?, ?)"
+        insert_traj_sql = "INSERT INTO Trajectory VALUES (?, ?, ?, ?, ?, ?, ?)"
+        vehicle_ids = set()
+        itcur = dbconxn.cursor()
+        for row in self.rows:
+            vid = int(self.column_val_in_row(row, "vehicle_id"))
+            if vid not in vehicle_ids:
+                ivcur = dbconxn.cursor()
+                veh_args = (
+                    vid,
+                    int(self.column_val_in_row(row, "type")),
+                    float(self.column_val_in_row(row, "length")) * self.scale,
+                    float(self.column_val_in_row(row, "width")) * self.scale,
+                )
+                ivcur.execute(insert_vehicle_sql, veh_args)
+                ivcur.close()
+                dbconxn.commit()
+                vehicle_ids.add(vid)
+            traj_args = (
+                vid,
+                # time units are in milliseconds for both NGSIM and Interaction datasets, convert to secs
+                round(float(self.column_val_in_row(row, "sim_time")) / 1000, 3),
+                float(self.column_val_in_row(row, "position_x")) * self.scale,
+                float(self.column_val_in_row(row, "position_y")) * self.scale,
+                float(self.column_val_in_row(row, "heading_rad")),
+                float(self.column_val_in_row(row, "speed")) * self.scale,
+                self.column_val_in_row(row, "lane_id"),
+            )
+            if not any(a is not None and np.isnan(a) for a in traj_args):
```

Comment:
I think the presence of `isnan` would be a value useful to warn about rather than silently skipping.

Reply:
yeah, in general I agree.  Unfortunately, the rolling window thing method used to calculate the moving averages for `position_x` and `position_y` leaves 1-window-width of NaNs at the end.  This was the easiest way to ignore those.  I'll add a comment to explain that.